### PR TITLE
ENH: log info when reading non-compliant EDF/EDF+ files

### LINKED ISSUE #11602: EDF / EDF+ header I/O
### Description of the problem

The header I/O code in [`mne.io.edf._read_edf_header()`](https://github.com/mne-tools/mne-python/blob/main/mne/io/edf/edf.py#L635) does not look entirely correct to me.

### Patient birthdate in 'local patient identification' field

According to [2.1.3. Additional specifications in EDF+](https://www.edfplus.info/specs/edfplus.html#additionalspecs):
> birthdate in dd-MMM-yyyy format using the English 3-character abbreviations of the month in capitals.

Therefore, this code is incorrect:
https://github.com/mne-tools/mne-python/blob/84259a673caafc3dfb37e07e9d73a3593c5c6acf/mne/io/edf/edf.py#L651
Indeed, `%b` stands for _Month as locale’s abbreviated name_, so it doesn't fit the _English 3-character abbreviations of the month_. It is perfectly OK to support locale month names for leniency and backwards compatibility, but the code should mainly support `JAN`, `FEB`, `MAR` and so on.

### Check for 'Startdate' in 'local recording identification'

According to [2.1.3. Additional specifications in EDF+](https://www.edfplus.info/specs/edfplus.html#additionalspecs):
> 4. The 'local recording identification' field must start with the subfields (subfields do not contain, but are separated by, spaces):
>    - The text 'Startdate'.

The following code should check for `Startdate`:
https://github.com/mne-tools/mne-python/blob/84259a673caafc3dfb37e07e9d73a3593c5c6acf/mne/io/edf/edf.py#L662

Typically:
```python
        if len(rec_info) == 5 and rec_info[0] == 'Startdate':
```
The code could only emit a warning for leniency and backwards compatibility.

### Startdate in 'local recording identification' field

Same as the birthdate above. According to [2.1.3. Additional specifications in EDF+](https://www.edfplus.info/specs/edfplus.html#additionalspecs):
> The startdate itself in dd-MMM-yyyy format using the English 3-character abbreviations of the month in capitals.

Therefore, this code is incorrect:
https://github.com/mne-tools/mne-python/blob/84259a673caafc3dfb37e07e9d73a3593c5c6acf/mne/io/edf/edf.py#L664
Indeed, `%b` stands for _Month as locale’s abbreviated name_, so it doesn't fit the _English 3-character abbreviations of the month_. It is perfectly OK to support locale month names for leniency and backwards compatibility, but the code should mainly support `JAN`, `FEB`, `MAR` and so on.

### Steps to reproduce

```Python
Read the code:
https://github.com/mne-tools/mne-python/blob/84259a673caafc3dfb37e07e9d73a3593c5c6acf/mne/io/edf/edf.py#L643-L672
```

### Link to data

_No response_

### Expected results

Follow the EDF+ specification, with appropriate forgiveness if needed.

### Actual results

Not compliant with the EDF+ specification.

### Additional information

[EDF and EDF+ specifications](https://www.edfplus.info/specs/index.html)

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
#### Reference issue

Fixes #11602.

#### What does this implement/fix?

Just log info when reading an EDF that cannot be interpredted as EDF+.

Emit warnings when the file is most probably an EDF+ file but is not EDF+-compliant.

In any case, do not make this an error, both for backwards compatibility and forgiveness in case we need to read 3rd party files.

#### Additional information

[EDF and EDF+ specifications](https://www.edfplus.info/specs/index.html)

### CODE REVIEW

File: mne/io/edf/edf.py [hunk 1]
```diff
@@ -638,7 +638,10 @@ def _read_edf_header(fname, exclude, infer_types, include=None):
 
     with open(fname, 'rb') as fid:
 
-        fid.read(8)  # version (unused here)
+        version = fid.read(8).decode('latin-1').rstrip()
```
Comment (cbrnr) [inline mne/io/edf/edf.py:641]: According to the specs, this must be ASCII encoded. I know we've already sprinkled in Latin-1, because it allows us to read slightly non-standard files, but at least in the version I think it should be ASCII.

File: mne/io/besa/besa.py [hunk 2]
```diff
@@ -86,13 +86,13 @@ def _read_evoked_besa_avr(fname, verbose):
     if 'Npts' in fields:
         fields['Npts'] = int(fields['Npts'])
         if fields['Npts'] != data.shape[1]:
-            logger.warn(f'The size of the data matrix ({data.shape}) does not '
-                        f'match the "Npts" field ({fields["Npts"]}).')
+            logger.warning(f'The size of the data matrix ({data.shape}) does '
+                           f'not match the "Npts" field ({fields["Npts"]}).')
```
Comment (larsoner) [inline mne/io/besa/besa.py:90]: We shouldn't use `logger.warning` but rather our own `warn` from `mne.utils`. It does some magic based on the logging level and also sets `stacklevel` etc.

Comment (DimitriPapadopoulos) [reply mne/io/besa/besa.py:90]: I copied `logger.warning` from other parts of the code base. Should I change to `warn` everywhere?

Comment (larsoner) [reply mne/io/besa/besa.py:90]: Yes I think so

Comment (DimitriPapadopoulos) [reply mne/io/besa/besa.py]: Done – except in a few places where comments state it shouldn't be changed.

**Comment by larsoner** [comment]
A lot of tests are failing and need to be fixed.

1. We treat warnings as errors, so these need to be caught or ignored now, or turned into `logger.info` if they seem unimportant.
2. There are a lot of `E   UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128)` as well. These could be `logger.info` or `warn`s, but based on the prevalence I'm guessing info?

**Comment by DimitriPapadopoulos** [comment]
The `UnicodeDecodeError: 'ascii' codec can't decode byte 0xff in position 0: ordinal not in range(128)` message should not occur when reading compliant EDF/EDF+ files. Besides, I don't see `0xff` anywhere in the header record of [`mne/io/edf/tests/data/test.edf`](https://github.com/mne-tools/mne-python/blob/main/mne/io/edf/tests/test_edf.py):
```
$ xxd /volatile/src/mne-python/mne/io/edf/tests/data/test.edf
00000000: 3020 2020 2020 2020 5820 5820 5820 5820  0       X X X X 
00000010: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000020: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000030: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000040: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000050: 2020 2020 2020 2020 5374 6172 7464 6174          Startdat
00000060: 6520 3239 2d41 5052 2d32 3031 3420 5820  e 29-APR-2014 X 
00000070: 5820 5820 2020 2020 2020 2020 2020 2020  X X             
00000080: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000090: 2020 2020 2020 2020 2020 2020 2020 2020                  
000000a0: 2020 2020 2020 2020 3239 2e30 342e 3134          29.04.14
000000b0: 3232 2e31 392e 3434 3336 3039 3620 2020  22.19.4436096   
000000c0: 4544 462b 4320 2020 2020 2020 2020 2020  EDF+C           
000000d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
000000e0: 2020 2020 2020 2020 2020 2020 3620 2020              6   
000000f0: 2020 2020 3120 2020 2020 2020 3134 3020      1       140 
00000100: 4131 2020 2020 2020 2020 2020 2020 2020  A1              
...
```
Let me investigate and fix this first.

**Comment by DimitriPapadopoulos** [comment]
Actually, the EDF file [mne/io/edf/tests/data/test.edf](https://github.com/mne-tools/mne-python/blob/main/mne/io/edf/tests/test_edf.py) is **not** EDF+-compliant. The 'local patient identification' field starts with `X X X X`:
```
$ xxd /volatile/src/mne-python/mne/io/edf/tests/data/test.edf
00000000: 3020 2020 2020 2020 5820 5820 5820 5820  0       X X X X 
00000010: 2020 2020 2020 2020 2020 2020 2020 2020                  
...
```
 I suspect this is the result of running [EDF De-identifier](https://prerau.bwh.harvard.edu/edf-de-identification-tool/) or a similar tool on the original EDF files. Unfortunately, [EDF De-identifier](https://prerau.bwh.harvard.edu/edf-de-identification-tool/) overwrites the 'local patient identification' field with `X X X X` and the 'local recording identification' field with `Startdate X X X X`, which is **not** EDF+-compliant. In the case of [mne/io/edf/tests/data/test.edf](https://github.com/mne-tools/mne-python/blob/main/mne/io/edf/tests/test_edf.py), the 'local patient identification' field starts with `X X X X` which is not compliant, but the 'local recording identification' starts with `Startdate 29-APR-2014 X X X`, which is compliant – the startdate has not been overwritten for de-identification purposes..

I'm not sure how to best handle this. I would like to emit a warning for `01-MEI-2000`, which really should be `01-MAY-2000` and is the result of using an incorrect locale, but probably would like to only write info to the logger for `X`, which is most certainly the result of a de-identification effort.

Perhaps I'll just write use `logger.info()` instead of `warn()` for now. It is a bit too late to flag compliance issues: this will break backwards compatibility.

**Comment by larsoner** [comment]
For `warn` vs `logger.info`, neither will break stuff for users. They see a message more often with `warn` than `logger.info` but code won't break. Our tests *as developers* break because we treat warnings as errors.

So at the end of the day it's a question of how visible we want these messages to be. To me if there is something that indicates a problem with the *data* -- like the expected number of samples not matching -- we should `warn` (more visible). If it's a problem with the *metadata* -- like recording date, subject information, etc. -- we should probably just `logger.info` (less visible).

**Comment by cbrnr** [comment]
I agree, we are not the EDF police, and I like the fact that we at least try to read some non-standard files (where other tools just fail).

**Comment by DimitriPapadopoulos** [comment]
I doubt we should (currently) warn about non-compliant EDF+ files. I would argue the MNE code should be even more forgiving.
- In the very first pseudnymised EDF+ files I have received in a new project, the 'recording identification' field contains only 2 items as in`Startdate 22-MEI-2019`, instead of the 5 expected items, as in `Startdate 22-MAY-2019 X X X`. I would like to be able to read the startdate even in this case (minus the locale error) – MNE currently fails to read that date.
- Another de-identifying tool, [edf-anonymize](https://archive.physionet.org/physiotools/edf-anonymize/), uses quotes to allow spaces in names in the 'patient identification' field. Should we be able to read such files? Perhaps not given that the EDF+ specification clearly states that “_subfields do not contain, but are separated by, spaces_”.
  > **edf-anonymize foo.edf anonymous.edf "Arthur Dent"**
  > As above, but the patient's name is replaced by Arthur Dent. The quotation marks are needed if the name contains any spaces, as in this example. Names longer than 80 characters are truncated. 
- W should accept `X` in the sex subfield, in addition to `F` and `M`, because again that's what pseudonymisation tools do.
- The compliance issues that I detect are actually created by other tools out there, such as [EDF De-identifier](https://prerau.bwh.harvard.edu/edf-de-identification-tool/). I would recommend first fixing those other tools, then changing the log level from info to warn in MNE.
 - Given the level of non-compliance in EDF files I have seen so far, which I understand is the norm, I suspect other tools are as much or more forgiving too.
 - Yet logging at info level might helps understand what MNE does under the hood, for example in my case detect that MNE is unable to read a 'recording identification' field which contains only two items.
 - Perhaps warnings should be kept for cases where MNE is not forgiving and is unable to make sense of the data.

**Comment by DimitriPapadopoulos** [comment]
I defer changing `logger.warning` to `warn` to another PR, as such a change is irrelevant for EDF I/O.

**Comment by cbrnr** [comment]
@DimitriPapadopoulos there are two relevant recent posts on our forum:

- https://mne.discourse.group/t/unable-to-read-edf-file/4569/11
- https://mne.discourse.group/t/error-reading-raw-bdf-data/4614/9

Since you are already working on trying to read more non-standard files, maybe you could also take a look at the files mentioned there? It seems these are also issues with invalid dates.

**Comment by DimitriPapadopoulos** [comment]
Let me see:

#### https://mne.discourse.group/t/unable-to-read-edf-file/4569/11

As you write over there, it is hard to tell without seeing the actual file. Yet, it looks like this is about the **“startdate” field** of the header record, in **`dd.mm.yy`** format, not the **“startdate” subfield** of the “recording identification” field, in **`dd-MMM-yyyy`** format. 

This looks like a serious error though, I don't see how we could have `'="text/c` instead of `dd.mm.yy` in a half-decent EDF file. This looks like an HTTP header downloaded as an EDF file, or something similar. Probably a hopeless case. 

We could of course catch the error in the date reading code, and fall back to `None`, which is already the case later on when building the final datetime:
https://github.com/mne-tools/mne-python/blob/3b4930abd32c00dc71da6c4a3e56a850a34ba185/mne/io/edf/edf.py#L687-L693
However, I suspect the specific file is rubbish any way.

#### https://mne.discourse.group/t/error-reading-raw-bdf-data/4614/9

This refers to the above issue.

We have additional information though, pointing to the [CHB-MIT Scalp EEG Database](https://physionet.org/content/chbmit/1.0.0/). i'll try to download the database and have a look at EDF files.

**Comment by cbrnr** [comment]
Thanks @DimitriPapadopoulos! I agree that without the actual file, we cannot debug. I've already asked the person to point us to the file from the CHB-MIT database that gives an error. The BDF file you mention is a different file, but should have a similar problem.

**Comment by DimitriPapadopoulos** [comment]
#### https://mne.discourse.group/t/error-reading-raw-bdf-data/4614

This post refers to the BDF file `sub-002/ses-01/eeg/sub-002_ses-01_task-meditation_eeg.bdf` from the [EEG meditation study](https://openneuro.org/datasets/ds001787/versions/1.0.3). Here is a dump of the 256 bytes of the header record of the `sub-002/ses-01/eeg/sub-002_ses-01_task-meditation_eeg.bdf`  file I have downloaded:
```
$ xxd sub-002/ses-01/eeg/sub-002_ses-01_task-meditation_eeg.bdf
00000000: ff42 494f 5345 4d49 2020 2020 2020 2020  .BIOSEMI        
00000010: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000020: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000030: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000040: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000050: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000060: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000070: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000080: 2020 2020 2020 2020 2020 2020 2020 2020                  
00000090: 2020 2020 2020 2020 2020 2020 2020 2020                  
000000a0: 2020 2020 2020 2020 3132 2e30 342e 3134          12.04.14
000000b0: 3136 2e30 382e 3031 3230 3733 3620 2020  16.08.0120736   
000000c0: 3234 4249 5420 2020 2020 2020 2020 2020  24BIT           
000000d0: 2020 2020 2020 2020 2020 2020 2020 2020                  
000000e0: 2020 2020 2020 2020 2020 2020 3237 3131              2711
000000f0: 2020 2020 3120 2020 2020 2020 3830 2020      1       80  
...
$ 
```
EDF+ and BDF files share the same header record format (256 bytes), which is why we use the same `_read_edf_header()` function to read the header record of both formats:
https://github.com/mne-tools/mne-python/blob/3b4930abd32c00dc71da6c4a3e56a850a34ba185/mne/io/edf/edf.py#L634-L635
However, the first 8 bytes differ in BDF files, `<0xff>BIOSEMI` in BDF instead of `0        ` in EDF. This is indeed part of the small differences between EDF and BDF documented in [BDF+](https://www.teuniz.net/edfbrowser/bdfplus%20format%20description.html):
> The version field must contain the following 8 octets: 0xff, 0x42, 0x49, 0x4f, 0x53, 0x45, 0x4d, 0x49

Ouch! I will need to adjust my changes to `_read_edf_header()` to allow for BDF files. 

Any way, I cannot find `;\n  posi` in the `sub-002/ses-01/eeg/sub-002_ses-01_task-meditation_eeg.bdf` BDF file I have downloaded. The value of the “startdate” field is perfectly valid: `12.04.14`. In short, I cannot explain the error message. It feels like this BDF file has somehow been corrupted in this post.

**Comment by cbrnr** [comment]
Thanks for investigating @DimitriPapadopoulos! Unless they make the file available there's nothing we can do.

**Comment by cbrnr** [comment]
There are some additional differences between EDF and BDF, which are nicely summarized here:

https://www.biosemi.com/faq/file_format.htm

**Comment by DimitriPapadopoulos** [comment]
#### https://mne.discourse.group/t/error-reading-raw-bdf-data/4614/5

We do not have access to this BDF file, but the value of the “startdate” field is `\x00\x00\x00\x01\x00\x00\x00\x04` instead of `dd.mm.yy`. While I could get MNE to skip such rubbish and set the resulting date to `None`, I suspect the rest of the file is corrupted too – chances are any data we manage read is nonsensical.

**Comment by cbrnr** [comment]
I agree, let's not try to read too many malformed EDF files. The way we do it now has been a good compromise between adhering to the standard but still being able to import data when some header fields are not quite what would be expected.

**Comment by DimitriPapadopoulos** [comment]
I could still make the code even more permissive, and accept EDF+ or BDF files with:
* less or more than 4 subfields in the “patient identification” field,
* less than 5 subfields in the “record identification” field.

However, this would require much more complex code, and obviously additional testing with files that are not compliant in diverse ways. Let's leave it aside for now.

I may implement this later, it if we cannot change the “record identification”  field from `Stardate 22-MAY-2019` to `Stardate 22-MAY-2019 X X X` in my current project. Chances are we will have to change to `Stardate X X X X`, `Stardate 01-JAN-1970 X X X` or something similar any way, to properly de-identify the files.

**Comment by cbrnr** [comment]
Agreed, let's focus only on logging existing checks in this PR (and the new version check is fine, of course). Re anonymizing data, the proper way to do that would be to insert some valid date and not `X X X X`, at least if you want a valid EDF file.

**Comment by larsoner** [comment]
@Teuniz these are all great points but would be a big change so I'll move that to a separate issue for now.

@DimitriPapadopoulos are you happy with the version now? If so @cbrnr feel free to review and merge (following whatever our current practice/philosophy is, more or less) and we can discuss being more strict separately.

**Comment by DimitriPapadopoulos** [comment]
@Teuniz With your much appreciated help, I Just noticed in the [EDF+ specification](https://www.edfplus.info/specs/edfplus.html):
> Subfields whose contents are unknown, not applicable or must be made anonymous are replaced by a single character 'X'

About the rest, we improve the situation by logging new info messages when reading EDF+ files. If others agree, I am 100% for changing them into warnings, perhaps it's best done before a minor release (1.4.0 in the case of MNE), not a patch release (1.3.2 in the case of MNE).

@larsoner I need to avoid logging messages in case we have `X` in de-identified files. These are valid EDF+ files.

**Comment by DimitriPapadopoulos** [comment]
@Teuniz About the “consensus to use 01-01-1985 00:00:00” in the “startdate” field of the header record, would you have a link to add to the source code comments, for future reference?

**Comment by Teuniz** [comment]
> @Teuniz With your much appreciated help, I Just noticed in the [EDF+ specification](https://www.edfplus.info/specs/edfplus.html):
> 
> > Subfields whose contents are unknown, not applicable or must be made anonymous are replaced by a single character 'X'
> 

That's why I wrote "The string "Startdate X X X X " in the recording field is valid in EDF+..." before.  
Or do you mean something else?

**Comment by Teuniz** [comment]
> @Teuniz About the “consensus to use 01-01-1985 00:00:00” in the “startdate” field of the header record, would you have a link to add to the source code comments, for future reference?

You can use "01.01.8500.00.00” at offset 0xA8 AND use "Startdate X X X X " at offset 0x58.

**Comment by Teuniz** [comment]
> I understand that, but a link to a page documenting this consensus would be great. Don't get me wrong, I totally trust you. A link would be useful for documentation purposes, not to convince me.

I'm not aware of any link/page. It's also not part of the EDF specification. It's just something I discussed in  
the past with other people and we always ended with this solution. Simply, because EDF doesn't allow to  
omit the startdate and/or starttime and the range is limited (100 years). So, the most sensible solution was to  
use a date and time as far as possible into the past which is 01-01-1985 & 00:00:00. This is the so called clipping  
date in EDF.

**Comment by DimitriPapadopoulos** [comment]
@ cbrnr I think I'm done. The new code does not emit new warnings, but it logs info level events whenever an EDF+ file is not compliant. I would be truly grateful if you could at some point run the new code on the public (or not) databases you use, as you had suggested, with info level logging on.
* This will help estimate the ratio of non-compliant files, and the kind of non-compliances out there. In turn, this could help change (some) info level events into actual warnings in the future.
* This will also help verify the new code doesn't do anything wrong on lots of EDF/EDF+/BDF files.

**Comment by cbrnr** [comment]
CirrusCI fails because one test modifies the logging level – can you make sure that this is unrelated? Furthermore, I'd prefer if you used f-strings instead of the .format() method if that's possible.

I've tested my local EDF and BDF files, and I get several messages related to "patient identification" and "recording identification" from this PR. I assume that people don't bother filling in these fields. Do you know how to set these fields to generic values that are compliant with EDF?

File: mne/io/edf/edf.py [hunk 3]
```diff
@@ -646,29 +649,61 @@ def _read_edf_header(fname, exclude, infer_types, include=None):
...
+                                    'record, which is not compliant with '
+                                    'EDF+, in file {}.'
+                                    .format(id_info[2], fname))
+                        birthdate = "X"
                 patient['sex'] = id_info[1]
                 patient['birthday'] = birthdate
                 patient['name'] = id_info[3]
+            else:
+                logger.info('Skipping the "patient identification" field '
+                            'of the  header record, which is not compliant '
```
Comment (cbrnr) [inline mne/io/edf/edf.py:670]:
```suggestion
                            'of the header record, which is not compliant '
```

File: mne/io/edf/edf.py [hunk 4]
```diff
@@ -646,29 +649,61 @@ def _read_edf_header(fname, exclude, infer_types, include=None):
...
         rec_info = fid.read(80).decode('latin-1').rstrip().split(' ')
         valid_startdate = False
         if len(rec_info) == 5:
-            try:
-                startdate = datetime.strptime(rec_info[1], "%d-%b-%Y")
-            except ValueError:
-                startdate = "X"
+            if rec_info[0] != 'Startdate':
+                logger.info('The "recording identification" field of the '
+                            'header record  begins with "{}", '
```
Comment (cbrnr) [inline mne/io/edf/edf.py:681]:
```suggestion
                            'header record begins with "{}", '
```

File: mne/io/edf/edf.py [hunk 5]
```diff
@@ -646,29 +649,61 @@ def _read_edf_header(fname, exclude, infer_types, include=None):
...
+                    startdate = "X"
+                else:
+                    valid_startdate = True
             meas_id['startdate'] = startdate
             meas_id['study_id'] = rec_info[2]
             meas_id['technician'] = rec_info[3]
             meas_id['equipment'] = rec_info[4]
+        elif len(rec_info):
+            logger.info('Skipping the "recording identification" field '
+                        'of the  header record, '
```
Comment (cbrnr) [inline mne/io/edf/edf.py:704]:
```suggestion
                        'of the header record, '
```

File: mne/io/edf/edf.py [hunk 6]
```diff
@@ -646,29 +649,61 @@ def _read_edf_header(fname, exclude, infer_types, include=None):
...
+                        'which is not EDF+-copmliant, in file {}.'
```
Comment (cbrnr) [inline mne/io/edf/edf.py:705]:
```suggestion
                        'which is not compliant with EDF+, in file {}.'
```

**Comment by DimitriPapadopoulos** [comment]
I prefer f-strings too, it's just that the rest of the file is using format. I'll fix that.

And perhaps change everything to f-strings in another PR?

**Comment by DimitriPapadopoulos** [comment]
Actually, it's already f-strings everywhere in this file. My wrong.

About the messages you get, I might need to be less aggressive. Are the message about skipping "patient identification" and "recording identification" **fields** altogether, or about skipping one of their **date subfields**?

**Comment by cbrnr** [comment]
The messages are about skipping the entire field, e.g.

```
Skipping the "recording identification" field of the  header record, which is not EDF+-copmliant, in file...
Skipping the "patient identification" field of the  header record, which is not compliant with EDF+, in file...
```

**Comment by cbrnr** [comment]
Just out of curiosity, are you force-pushing because you rebased your branch?

**Comment by DimitriPapadopoulos** [comment]
(Yes, I rebased + an accidental copy/paste in my text editor)

I see. These are probably EDF files, not EDF+ files.

The EDF standard does not impose any constraint over the 80 bytes of the "patient identification" and "recording identification" fields. They just need to be ASCII characters.

The EDF+ standard does iimpose constraints: 4 subfields separated by spaces in the "patient identification" field, and 5 subfields separated by spaces in the "recording identification" field.

I may have misunderstood something, but I think we cannot distinguish between EDF and EDF+ files. Therefore, the heuristic is that if we have 4 words separated by spaces in the "patient identification" field, or 5 words separated by spaces in the "recording identification" field, we assume we are reading an EDF+ file. In that case we log an info message if the fields cannot be interpredted according to the EDF+ specs. However, if it's a simple EDF file, no message should be logged. Except we cannot guess the type of the file.

**Comment by cbrnr** [comment]
> (Yes, I rebased + an accidental copy/paste in my text editor)

OK! I was not sure if you know that we squash and merge, so you don't have to maintain clean commits.

Regarding EDF vs. EDF+, I think the first reserved header field should be different for these two versions: https://www.edfplus.info/specs/edfplus.html#header

**Comment by Teuniz** [comment]
Indeed. The only correct and reliable way to check if an EDF file is of type EDF or EDF+, is by examining the reserved  
field in the header (44 bytes).  
If it's completely empty (only spaces), it's EDF.  
If it starts with "EDF+" (without the quotes) and the remaining bytes are spaces, it's EDF+.  
Every other possibility renders the file invalid.  
In case of EDF, any EDF+ style formatting must be ignored, the file needs to be interpreted like old style EDF.

**Comment by DimitriPapadopoulos** [comment]
Back to _Draft_.

The current code is perfect for a best effort approach to extract data from the "patient identification" and "recording identification" fields despite compliance issues. It won't do for properly logging compliance issues. We need additional branching to handle EDF, EDF+, BDF  and BDF+ file types separately, when needed.

**Comment by Teuniz** [comment]
Sorry for barging in again but may I ask what it is exatly what you want to try to achieve?
Because it looks like you are trying to reinvent the wheel by writing your own EDF compatibility checker
while there are already two good EDF compatibility checkers.

Let's forget for a moment about EDFbrowser but focus on Polyman instead.
Polyman (1, 2) is a free (but closed source) EDF viewer and annotator, written by Marco Roessen (3)
under direct supervision of Bob Kemp (4, 5, 6, 7) who was (he's retired now) the lead inventor of
the EDF and EDF+ format.
Polyman can be downloaded and used without registration and there are no strings attached.
Polyman has an EDF(+) compatibility checker.

What about this idea:
Use the EDFreader() (8) and if it raises an exception,
simply inform the user that the file is non-compliant and that, for more details about the problem and a
possible fix, they can use Polyman or EDFbrowser. No need to reinvent the wheel.

1) https://www.edfplus.info/downloads/

2) https://sites.google.com/view/diegoalvarezestevez/projects/polyman

3) https://nl.linkedin.com/in/marco-roessen-a0909430

4) https://www.researchgate.net/profile/Bob-Kemp

5) https://www.sciencedirect.com/science/article/abs/pii/0013469492900097

6) https://pubmed.ncbi.nlm.nih.gov/12948806/

7) https://physionet.org/content/sleep-edf/1.0.0/

8) https://pypi.org/project/EDFlib-Python/

**Comment by cbrnr** [comment]
That's a valid question @Teuniz!

First, I would like to avoid regressions, meaning that I don't want to read fewer EDF files than before. Currently, we do read some non-standard EDF files, because I think it is a bit silly to prevent loading data just because one non-essential header field is not following the standard.

Second, and this is related to the first point, being super strict is only viable if we provide an option to "fix" a non-standard EDF file out of the box. IMO, this option must be (1) open source, (2) reasonably fast, (3) callable from Python, and (4) pure Python if possible. None of the tools you mention fulfill all criteria, so we're stuck with what we have right now.

Finally, I agree that it is not clear what we would like to do. I don't think re-inventing the wheel by upgrading our code to be a full compatibility checker is a reasonable goal. I'd rather stick with what we have, because in my experience we can load >99% of EDF files out there. The problematic cases are always related to some non-standard dates and/or non-standard character encoding in some header fields. I'd rather fix those problems on an individual basis, meaning that if they deviate too far from the standard I'd happily point people to go fix their files with one of the tools you mention. The goal of this PR was to just issue an info (or a warning) if we encounter a non-standard field that we have been fixing silently before.

In summary, maybe there really isn't much to do right now. Let's focus on concrete use cases that currently don't work, but that we think we can and should still be able to read.

**Comment by cbrnr** [comment]
But we very likely skip a lot of fields silently, because we have never needed it. I'm fine with adding checks for more individual fields, but we have to be aware that we will still not check if a file is 100% EDF compatible. We will and should probably never do this, although I think it won't be that complicated. So yes, if we are explicitly skipping or not validating a field, we can log it, but I don't want that to prevent us from loading the file.

And yes, I still think a strict/loose mode for the reader would be a nice addition, but let's discuss this in #11611.

So for this PR, how would you like to proceed @DimitriPapadopoulos?

**Comment by Teuniz** [comment]
>  I started with [pyEDFlib](https://github.com/holgern/pyedflib). It looked like a good choice. It appears to be the most popular Python module for EDF+ I/O. It is based on Teunis' [EDFlib](https://gitlab.com/Teuniz/EDFlib) which is the reference open source library for EDF+ I/O as far as I can see. The only issue was the lack of detailed error messages, I find messages such as “_the file is not EDF(+) or BDF(+) compliant_” are not detailed enough.

Pay attention, [pyEDFlib](https://github.com/holgern/pyedflib) is actually a fork of EDFlib. They applied their own patches because they don't like my policy regarding non-compliant files. So, using their fork to check for compatibility is a bad idea.

> Then I tried [pyedf](https://github.com/bids-standard/pyedf). I gave up rapidly because the main branch is stuck with Python 2 for reasons that do not make real sense to me.

Did you try [EDFlib for Python](https://pypi.org/project/EDFlib-Python/)?

Finally, you can use this extremely fast [EDF de-identifier](https://www.teuniz.net/edf_deidentifier/) in dry-run mode as
an EDF compatibility checker:

`edf-deidentifier -f filename -p -d`

Have a look at the README for an explanation.

**Comment by Teuniz** [comment]
> The only issue was the lack of detailed error messages, I find messages such as “_the file is not EDF(+) or BDF(+) compliant_” are not detailed enough.

The reason for that is that I don't want to maintain duplicate code for something that doesn't fix the problem anyway.
